### PR TITLE
Download geoJSONs from Louvre- file per object type, provider, company and date

### PR DESCRIPTION
**Pull request by rotemgabbai**

### CODE REVIEW

File: Download_geoJSONs_from_Louvre.py [hunk 1]
```diff
@@ -19,21 +20,39 @@
...
+            groups[obj['observed_start'].strftime('%Y%m%d') + obj['observed_end'].strftime('%Y%m%d') + obj['company']].append(obj)
+            grouped_list = list(groups.values())
 
-    for obj in object_list:
-        fileName = obj['company'] + '_' + obj['_id']
-        writeToJSONFile(download_directory, fileName, obj)
-        
\ No newline at end of file
+        for gl in grouped_list:
+            if (gl[0]['observed_start'] == gl[0]['observed_start']):
```
Comment (mzedayda) [inline Download_geoJSONs_from_Louvre.py]: Always true

File: Download_geoJSONs_from_Louvre.py [hunk 2]
```diff
@@ -19,21 +20,39 @@
...
+                fileName = gl[0]['company'] + '_' + object_type + '_' + gl[0]['observed_start'].strftime('%Y%m%d') 
+            else:           
+                fileName = gl[0]['company'] + '_' + object_type + '_' + gl[0]['observed_start'].strftime('%Y%m%d') + gl[0]['observed_end'].strftime('%Y%m%d')
```
Comment (mzedayda) [inline Download_geoJSONs_from_Louvre.py]: missing _ between dates

File: Download_geoJSONs_from_Louvre.py [hunk 3]
```diff
@@ -19,21 +20,39 @@
...
+            obj['properties'] = {}            
+            groups[obj['observed_start'].strftime('%Y%m%d') + obj['observed_end'].strftime('%Y%m%d') + obj['company']].append(obj)
+            grouped_list = list(groups.values())
 
-    for obj in object_list:
-        fileName = obj['company'] + '_' + obj['_id']
-        writeToJSONFile(download_directory, fileName, obj)
-        
\ No newline at end of file
+        for gl in grouped_list:
```
Comment (mzedayda) [inline Download_geoJSONs_from_Louvre.py]: g1?

File: Download_geoJSONs_from_Louvre.py [hunk 4]
```diff
@@ -19,21 +20,39 @@
...
         }
+
     object_list = list(db[object_type.lower()].find(query))
+    groups = defaultdict(list)
+
+    if len(object_list) > 0:
+        for obj in object_list:
+            obj['type'] = 'Feature'
+            obj['properties'] = {}            
+            groups[obj['observed_start'].strftime('%Y%m%d') + obj['observed_end'].strftime('%Y%m%d') + obj['company']].append(obj)
```
Comment (mzedayda) [inline Download_geoJSONs_from_Louvre.py]: too complicated, split to 2 lines

File: Test/testMongo.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+config = json.load(open(sys.argv[1]))
+download_directory = abspath(config['download_directory'])
+start_date = datetime.strptime(config['start_date'], '%Y-%m-%dT%H:%M:%S.000Z')
+end_date = datetime.strptime(config['end_date'], '%Y-%m-%dT%H:%M:%S.000Z')
+object_types = config['object_types']
+companies = config['companies']
+geometry = config['geometry']
+
+item1 = {
+        '_id' : 'test1',
```
[the remark this answers is not in the document]
Comment (taihak94) [reply Test/testMongo.py]: remove the extra  space before the `:` -> `"_id": 'test'`
do this for everything

File: Test/testMongo.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+                ]
+            ],
+            'type' : 'Polygon'
+        },
+        'observed_start' : datetime.strptime('2018-12-17T06:11:00.000Z','%Y-%m-%dT%H:%M:%S.000Z'),
+        'observed_end' : datetime.strptime('2018-12-21T06:11:00.000Z','%Y-%m-%dT%H:%M:%S.000Z')
+    }
+
+db.test.replace_one({ '_id': 'test1' }, item1, upsert = True)
+
```
Comment (taihak94) [inline Test/testMongo.py]: remove extra line

File: Test/testMongo.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+db.test.replace_one({ '_id': 'test2'}, item2, upsert = True)
+
+query = {
+            'observed_start': { '$lt': end_date },
+            'observed_end': { '$gt': start_date },
+            'company': { '$in': companies },
+            'geometry' : { '$geoIntersects': { '$geometry': geometry } }
+        }
+
+test_list = list(db.test.find(query))
```
Comment (taihak94) [inline Test/testMongo.py]: Add a comment explaining what you're trying to show in the test

File: Download_geoJSONs_from_Louvre.py [hunk 8]
```diff
@@ -19,21 +20,41 @@
...
 
 for object_type in object_types:
     query = {
-                'observed' : { '$gte' : start_date, '$lt': end_date },
-                'company' : { '$in': companies }, 
-                'geometry' : { '$geoIntersects': { '$geometry': geometry } }
+            'observed_start': { '$lt': end_date },
+            'observed_end': { '$gt': start_date },
+            'company': { '$in': companies }, 
+            'geometry' : { '$geoIntersects': { '$geometry': geometry } }
```
Comment (taihak94) [inline Download_geoJSONs_from_Louvre.py]: change this to a method that gets `start, end, company, geometry` and returns a query

File: Test/testMongo.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+check = (test_list == list(db.test.find({ '_id': 'test4' })))
+assert check == True, 'test failed'
```
Comment (taihak94) [inline Test/testMongo.py]: shouldn't contain `item2`

File: Test/testMongo.py [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+
+query = {
+            'observed_start': { '$lt': end_date },
+            'observed_end': { '$gt': start_date },
+            'company': { '$in': companies },
+            'geometry' : { '$geoIntersects': { '$geometry': geometry } }
+        }
+
+test_list = list(db.test.find(query))
+check = (test_list == list(db.test.find({ '_id': 'test4' })))
```
Comment (taihak94) [inline Test/testMongo.py]: `check = (not ('item2' in test_list))`

File: functions.py [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+from collections import defaultdict
+
+def datetime_handler(x):
+    if isinstance(x, datetime):
+        return x.isoformat()
+    raise TypeError("Unknown type")
+
+def create_query (start_date, end_date, companies, geometry):
+    query = {
+            'observed_start': { '$lt': end_date },
```
Comment (mzedayda) [inline functions.py]: Too much indent

**Comment by mzedayda** [comment]
Tests should be fixed:
* Move test data to separate file.
* Use unittest module, see [pyCoral](https://github.com/nsplt/pycoral) for example

File: functions.py [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+        'company': { '$in': companies }, 
+        'geometry' : { '$geoIntersects': { '$geometry': geometry } }
+    }
+    return query
+
+
+def writeToJSONFile (path, fileName, data):
+    filePathName = join(path, fileName + '.json')
+    file_data = {
+        "type": "FeatureCollection",
```
Comment (mzedayda) [inline functions.py]: Only ', not "

**Pull request merged**